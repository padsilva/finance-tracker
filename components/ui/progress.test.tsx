import { render } from "@testing-library/react";

import { Progress } from "./progress";

// Mock Radix UI Progress primitive
jest.mock("@radix-ui/react-progress", () => ({
  Root: ({
    children,
    className,
    ...props
  }: {
    children: React.ReactNode;
    className?: string;
  }) => (
    <div data-testid="progress-root" className={className} {...props}>
      {children}
    </div>
  ),
  Indicator: ({
    className,
    style,
    ...props
  }: {
    className?: string;
    style?: React.CSSProperties;
  }) => (
    <div
      data-testid="progress-indicator"
      className={className}
      style={style}
      {...props}
    />
  ),
}));

describe("Progress Component", () => {
  it("should render with default state (no value)", () => {
    const { container } = render(<Progress />);

    expect(container).toMatchSnapshot();
  });

  it("should render with different progress values", () => {
    const { container } = render(
      <>
        <Progress value={0} />
        <Progress value={25} />
        <Progress value={50} />
        <Progress value={75} />
        <Progress value={100} />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom className", () => {
    const { container } = render(
      <Progress className="custom-progress h-4" value={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should handle edge cases", () => {
    const { container } = render(
      <>
        <Progress value={-10} /> {/* Should clamp to 0 */}
        <Progress value={150} /> {/* Should clamp to 100 */}
        <Progress value={undefined} />
        <Progress value={null} />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with different sizes", () => {
    const { container } = render(
      <>
        <Progress className="h-1" value={50} />
        <Progress className="h-2" value={50} />
        <Progress className="h-3" value={50} />
        <Progress className="h-4" value={50} />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom colors", () => {
    const { container } = render(
      <>
        <Progress className="bg-red-200 [&>div]:bg-red-500" value={50} />
        <Progress className="bg-green-200 [&>div]:bg-green-500" value={50} />
        <Progress className="bg-blue-200 [&>div]:bg-blue-500" value={50} />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with aria attributes", () => {
    const { container } = render(
      <Progress
        value={75}
        aria-label="Download progress"
        aria-valuemin={0}
        aria-valuemax={100}
        aria-valuenow={75}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with data attributes", () => {
    const { container } = render(
      <Progress
        value={50}
        data-testid="custom-progress"
        data-state="loading"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render in different states", () => {
    const { container } = render(
      <>
        <Progress value={0} className="opacity-50" /> {/* Inactive */}
        <Progress value={100} className="opacity-100" /> {/* Complete */}
        <Progress value={50} className="animate-pulse" /> {/* Loading */}
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with custom animations", () => {
    const { container } = render(
      <>
        <Progress
          value={50}
          className="transition-all duration-500 ease-in-out"
        />
        <Progress
          value={50}
          className="animate-[progress_1s_ease-in-out_infinite]"
        />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with decorative elements", () => {
    const { container } = render(
      <div className="space-y-2">
        <div className="flex justify-between text-sm">
          <span>Progress</span>
          <span>50%</span>
        </div>
        <Progress value={50} />
        <div className="text-xs text-muted-foreground">Processing...</div>
      </div>,
    );

    expect(container).toMatchSnapshot();
  });
});
