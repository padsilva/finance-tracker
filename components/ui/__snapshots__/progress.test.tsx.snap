// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Component should handle edge cases 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-110%);"
    />
  </div>
   
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(--50%);"
    />
  </div>
   
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-100%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render in different states 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 opacity-50"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-100%);"
    />
  </div>
   
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 opacity-100"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-0%);"
    />
  </div>
   
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 animate-pulse"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
   
</div>
`;

exports[`Progress Component should render with aria attributes 1`] = `
<div>
  <div
    aria-label="Download progress"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="75"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-25%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with custom animations 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 transition-all duration-500 ease-in-out"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 animate-[progress_1s_ease-in-out_infinite]"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with custom className 1`] = `
<div>
  <div
    class="relative w-full overflow-hidden rounded-full bg-primary/20 custom-progress h-4"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with custom colors 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-red-200 [&>div]:bg-red-500"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-green-200 [&>div]:bg-green-500"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-blue-200 [&>div]:bg-blue-500"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with data attributes 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-state="loading"
    data-testid="custom-progress"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with decorative elements 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Progress
      </span>
      <span>
        50%
      </span>
    </div>
    <div
      class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
      data-testid="progress-root"
    >
      <div
        class="h-full w-full flex-1 bg-primary transition-all"
        data-testid="progress-indicator"
        style="transform: translateX(-50%);"
      />
    </div>
    <div
      class="text-xs text-muted-foreground"
    >
      Processing...
    </div>
  </div>
</div>
`;

exports[`Progress Component should render with default state (no value) 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with different progress values 1`] = `
<div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-100%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-75%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-25%);"
    />
  </div>
  <div
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-0%);"
    />
  </div>
</div>
`;

exports[`Progress Component should render with different sizes 1`] = `
<div>
  <div
    class="relative w-full overflow-hidden rounded-full bg-primary/20 h-1"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative w-full overflow-hidden rounded-full bg-primary/20 h-2"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative w-full overflow-hidden rounded-full bg-primary/20 h-3"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
  <div
    class="relative w-full overflow-hidden rounded-full bg-primary/20 h-4"
    data-testid="progress-root"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-testid="progress-indicator"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;
